### ISSUE TITLE
plugins/sources/git.py: Do not checkout submodules by default

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#783]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/783)
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 25, 2018, 19:39

## Background
We do not really want, as a default, to download things from internet the user have not specify in their .bst files

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 25, 2018, 19:40

mentioned in merge request !972

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 25, 2018, 20:01

mentioned in merge request freedesktop-sdk/freedesktop-sdk!722

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 27, 2018, 09:38

assigned to [[Gitlab user @jjardon]](https://gitlab.com/jjardon)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 29, 2018, 12:03

> We do not really want, as a default, to download things from internet the user have not specify in their .bst files

I think the fact that the `git` module has submodules at a given commit, is an indication that those modules are required to build.

I think the deciding factor to consider here (asides from this being a breaking change which is already quite bad, we've been making *far* too many of these this cycle and hope we can stop doing it **completely** in one cycle or two...) is what is more convenient for the user ?

Consider the average: for all of the projects we tend to build which have submodules, do project authors have to disable checking out of submodules more than they would have to enable checking out of submodules ?

Also consider the case where a project author tracks a `git` source, and the new ref now points to a commit which has submodules; it seems to me quite disingenuous for their build to fail as a result of tracking, just because they did not consider having to manually *enable* checking out of the new submodules.

It also seems to me the motivation here is mostly just an optimization, which is good to consider but secondary to having things *"just work out of the box"*.

In other words; BuildStream should provide an experience which *"just works"* as much as possible, while requiring the minimal amount of setup from the project author. If the project author wants to go ahead and *optimize* the build to exclude the download of a submodule which for some reason is unneeded for the build, then they can explicitly go ahead and do that... but at least by default things *"just work"*.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 29, 2018, 12:17

Honestly regardless of the comments above about why I think we currently *do* have the correct default, I'm very inclined to reject this solely based on the fact that it is a very invasively breaking change.

Literally any project which uses any git sources which have submodules would break as a result of this change - while I know for a fact that a known project like `gnome-build-meta` has *several* (probably even close to 10, definitely more than 5) elements which have submodules which the authors previously did not have to care about - it stands to reason that there is a very good chance this breaks *any* projects that we dont know about and track closely - it's a very unfriendly change.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 29, 2018, 13:56

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) ok, looking at https://gitlab.com/BuildStream/website/issues/24, I though some breaking changes were allowed in this cycle

If this is not the case, and even if I think this default behaviour is broken, would you be happy to change this MR to update the docs to warn about this?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @nanonyme]](https://gitlab.com/nanonyme) on Nov 29, 2018, 17:55

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) note that also devs from Gnome were shocked after realizing the default. It seems likely most Buildstream users will turn the feature off by default before 1.4 is out and set explicit submodules where necessary. Of course, if you can assume everyone has projects done this way, this won't really be such a breaking change anymore...

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 30, 2018, 07:06

We discussed this at length on IRC yesterday, the log of which [starts here](https://irclogs.baserock.org/buildstream/%23buildstream.2018-11-29.log.html#t2018-11-29T11:49:42)

I still firmly believe the default we have is the correct one, while I agree that users need to have more assistance and increased awareness around which of their modules happen to use, or start using a submodule.

There are some opinions on both sides of this, and as there is even a semantic to change the default project wide, there is no grounds to introduce a breaking change for this, so we will not.

For the sake of completeness, I will summarize my understanding of the arguments on either side:

* **Disable submodules by default**

  This proposal is largely driven by a module (I'm still not sure which one, gstreamer ?) which does something I consider to be rather fringe and unexpected; i.e. it does a conditional initialization of it's submodules *from it's build scripts* depending on what is found on the host system.

  I.e. it will build its own copy of something depending on whether a dependency exists.

  This fringe case, or set of cases is/are causing `freedesktop-sdk` maintainers frustration because they were not previously informed of the module even using any submodules; so their natural reaction is to presume that "Hey, I didn't ask for these submodules explicitly, so why am I using them ?"

  I still hold the firm opinion that the *not knowing* part of this is the issue, not the default behavior. And the not knowing part can be easily addressed in another way which does not introduce a breaking change and also adds further value to BuildStream.

* **Enable submodules by default**

  The normal mode of operation of a project which uses submodules is that the module code shares something with other git repositories and it is not conditionally enabled by build scripts.

  The defaults of BuildStream should always be optimized towards having things "just work" out of the box, once things are found to not work as a maintainer expects, then explicit settings can be added to individual `.bst` files in those cases, but the default case should "just work" for the majority.

  Changing the default of submodules to be disabled as opposed to the current default, is proposing that elements which inadvertently use submodules **break** the majority of the time, until the user figures out on their own that the module in question needs a submodule to be explicitly enabled.

  BuildStream does not have any way to know, or much less inform the user, that the reason for a build breaking after adding a new element or running `bst track`, is because the user forgot to go ahead and explicitly enable the submodule. The most likely scenario is that the user gets a broken build telling them `"No rule to build foo.c"`, and let them figure it out for themselves.


Both scenarios have their drawbacks, but the drawbacks of enabling submodules by default are much less common, and can be addressed with better messaging, while the drawbacks of changing them to be disabled by default make life inconvenient for a larger spectrum of cases, and cannot easily be addressed by other means (as stated, we cannot *know* that a build failed because of a submodule, that BuildStream was not told to expect).


### Better messaging

Throughout the above text I've been alluding to addressing the current frustrations with better messaging, I'll elaborate on what is meant by this here:

* The `git` plugin supports explicit mentions of the URLs and paths of submodules it is expected to use, this is important to ensure project relocatability (building the same project from different, relocated source URLs), and is mandatory in the case of using source mirroring fallback urls, both of which rely on [source aliasing](https://docs.buildstream.build/format_project.html#source-aliases) to work.

* Since there are features which cannot work without explicit aliasing of the submodule URLs, it stands to reason regardless that the user should be warned whenever a git source starts using a submodule that is not declared in the `.bst` file (which can happen when the source is added, or as a result of `bst track`).

* Of course if a specific module wants to explicitly ignore / disable usage of submodules, the said warning is irrelevant to tell the user, since the user is already explicitly ignoring the submodule portion of the git input.

* As with other warnings, these can always be configured as fatal, if a project wants to enforce that all `git` sources have the URLs specified, it can error out when they are not.

While this warning is already important to add regardless of this issue, in order to ensure proper aliasing of project input URLs, it will also have the side effect of raising awareness whenever a submodule inadvertently presents itself, leaving the user the choice of whether to explicitly ignore it or not - without damaging the default use cases of *"Things just working with a warning"*.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 30, 2018, 08:14

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) I think you are still missing the point. This issue is about control. The idea that having a look to my .bst files I know what exact elements are being built (and yeah, I want it to fail if some module tries to download something I have not specify in any if my .bst files)

The gstreamer case was an example, but a lot more can exist

About

> There are some opinions on both sides of this

sorry, but you are you are the only one I know thinking the current default is the good one

Now, if changing the default is totally not an option I think we should improve this situation with:
- Documentation
- warnings in the build logs (which maybe we can make fatal)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 30, 2018, 09:01

> [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) I think you are still missing the point. This issue is about control.

You have control regardless of the default, the default is just not what *you* expect it to be.

I maintain that in either case, the issue is about awareness of what is what, which is what the warning approach will provide.

> > There are some opinions on both sides of this
>
> sorry, but you are you are the only one I know thinking the current default is the good one

I think this is rather inflammatory, people don't tend to complain when things work. Your proposal clearly suggests a situation which causes things to not work by default; most of the time. As soon as we change things which cause things to *not work* by default, then the other side of the argument will be more apparent, and in the form of complaints.

> Now, if changing the default is totally not an option I think we should improve this situation with:

I do think that this breaking change is not an option, it is not a simple CLI invocation change that will be a temporary annoyance to a few users at upgrade time, it is a change that breaks virtually any project which uses submodules.

We need to be even more careful about breaking the projects we don't know about than the ones we do know about and can patch; even if the numbers of those might be small, we need to focus on inspiring confidence in those who don't openly communicate with us, and we need to put a stop to the breakage fiaso as soon as possible.

When we talk about moving plugins into separate repositories etc, we might have an opportunity at that time to provide a transition to an alternative API for the `git` plugin at which time this conversation could happen again more meaningfully, but for concrete reasons I think I have explained very clearly (as opposed to just claiming *"everyone doesnt expect this"*, which I think is rather hand wavy and not an argument backed up by facts), I will most probably remain in the *enabled by default* camp of that conversation, unless more real arguments based on fact present themselves later.

> * Documentation

This looks like simply a statement in the `git` plugin documentation front and center which says that *This plugin will automatically stage the submodules by default*, which is true. Perfectly acceptable one liner merge request I think.

> * warnings in the build logs (which maybe we can make fatal)

Yes, which I have already outlined and suggesting, and are useful in either case.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 30, 2018, 12:27

> I think this is rather inflammatory,

sorry, that was not my intention

I will submit a MR for the docs shortly

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Nov 30, 2018, 20:28

For what it's worth, in my case we needed to turn submodules off as well.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Dec 2, 2018, 21:05

mentioned in commit f4e2a04b6d02a1c3a06c8cc1003343bed0e532e6

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Dec 2, 2018, 21:06

mentioned in merge request !983

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Dec 3, 2018, 09:37

mentioned in commit efe1bbd8bd410e268115e6f54a0562bfa9922dec

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Dec 3, 2018, 11:14

mentioned in commit 9022a778827ad5fe6f39daa835aff6002dccf14c

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Dec 3, 2018, 14:38

unassigned [[Gitlab user @jjardon]](https://gitlab.com/jjardon)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 18, 2020, 14:11

mentioned in commit 4ecefebcc18547b422d401cf5556c9821d02934b

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 18, 2020, 15:31

mentioned in commit 3bcbcaff7363f47507003126fd81f0b4366f89cf

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 20, 2020, 15:57

mentioned in commit c9a1588f44abb71e93f584d88816dc30ded6ef40

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 20, 2020, 16:03

mentioned in commit 594b3dbc2fe952c7668ea2d1290856a1bbf9138a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 20, 2020, 22:36

mentioned in commit 7d37de3b5bd828a9439b939afcf237df2f601b6d

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 22, 2020, 17:37

mentioned in commit c89b5c7c5297845d6a5d2fe33b3290f85391ff64

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on May 1, 2020, 20:05

mentioned in commit 25125d38ee54ac79056e246f900477e06490dbcb
